text/0000-semantic-test-selectors.md

```diff
@@ -0,0 +1,700 @@
+select elements, we can open the door for further evolution.
+
+### Cleaner tests
+
+This is perhaps the most visceral motivation. Which test would you rather read?
+
+This one:
+
+```js
+await fillIn('.login-form .field-email', 'alice@example.com');
+await fillIn('.login-form .field-password', 'topsecret');
+await click('.submit-btn');
+```
+
+Or this one:
+
+```js
+await fillIn('Email', 'alice@example.com');
+await fillIn('Password', 'topsecret');
+await click('Log in');
+```
+
+## Detailed design
+
+We will provide semantic variants of `click` and `fillIn`. Their proposed
+behaviours are given below.
+
+### Semantic `click`
+
+```ts
+function click(label: string): Promise<void>
+```
+
+This helper finds a **link** or **button** whose **perceivable text** contains
+the string specified by `label`, or a **switch** whose **perceivable label**
+contains the string specified by `label`.
+
+If it finds exactly one matching element, it invokes the general-purpose
+`click` helper upon it.
+
+If it finds multiple matching elements, it throws:
+
+```
+Found multiple clickable elements with perceivable text '${label}' when calling
+`click('${label}')`. You may want to use a more descriptive label in the test
+and the template.
+```
+
+If it finds no matching element, it throws:
+
+```
+Clickable element with perceivable text '${label}' not found when calling `click('${label}')`.
+```
+
+### Semantic `fillIn`
+
+```ts
+function fillIn(label: string, value: string): Promise<void>
+```
+
+This helper finds an **input** whose **perceivable label** contains the string
+specified by `label`.
+
+If it finds a matching element, it invokes the general-purpose `fillIn` helper
+upon it with `value` as the second argument.
+
+If a label is found without a counterpart form control it throws:
+
+```
+Form control not found for label containing '${label}' when calling
+`fillIn('${label}', '${value}')`. You may want to check the `for` attribute of
+the label.
+```
+
+If no label is found, it throws:
+
+```
+Form control labelled '${label}' not found when calling `fillIn('${label}', '${value}')`.
+If the form control exists, you may want to label its using a label, title or aria-label.
+```
+
+### Definitions
+
+A **link** is one of the following:
+
+- `a`
+- `[role="link"]`
+- `[role="menuitem"]`
+
+A **button** is one of the following:
+
+- `button`
+- `input[type="button"]`
+- `input[type="reset"]`
+- `input[type="submit"]`
+- `[role="button"]`
+
+An **input** is one of the following:
+
+- `input`
+- `textarea`
+- `select`
+- `[role="slider"]`
+- `[role="spinbutton"]`
+- `[role="textbox"]`
+- `[contenteditable="true"]`
+
+A **switch** is one of the following:
+
+- `input[type="checkbox"]`
+- `input[type="radio"]`
+- `[role="checkbox"]`
+- `[role="option"]`
+- `[role="radio"]`
+
+A **form control** is an **input** or **switch**.
+
+The **perceivable text** of a **link**, **button**, or other element may be
```

Comment:
we should make it clear, that if you have all the attributes it will select one in a deterministic order that will be used as it's perceivable string and not the others

@MelSumner do you know where in the spec it describes this?

Reply:
I think what we’re after is the specification of [Text Alternative Computation](https://www.w3.org/TR/accname-1.1/#mapping_additional_nd_te)

Reply:
On which note: creating an implementation of that spec actually looks like a pretty fun exercise! (to me, at any rate)

Reply:
first pass at it :) need to build a proper test suite before improving https://github.com/tradegecko/ember-semantic-test-helpers/commit/91fd983896f60b0f8352258f895a3028835b71f5